### PR TITLE
BUG SparseDataFrame with dense Series (#19374)

### LINKED ISSUE #19374: SparseDataFrame with dense Series or unknown type
#### Code Sample, a copy-pastable example if possible

```python
import pandas

pandas.SparseDataFrame(pandas.Series())
```
#### Problem description

I've seen there are plans to SparseDataFrame. But reporting anyway.

The above code, creating a `SparseDataFrame` with a dense `Series` fails with a not very descriptive error:

```
---------------------------------------------------------------------------
UnboundLocalError                         Traceback (most recent call last)
<ipython-input-6-b9708ff05a7c> in <module>()
      1 import pandas
      2 
----> 3 pandas.SparseDataFrame(pandas.Series())

~/anaconda3/lib/python3.6/site-packages/pandas/core/sparse/frame.py in __init__(self, data, index, columns, default_kind, default_fill_value, dtype, copy)
    120                 mgr = mgr.astype(dtype)
    121 
--> 122         generic.NDFrame.__init__(self, mgr)
    123 
    124     @property

UnboundLocalError: local variable 'mgr' referenced before assignment
```

Any other type provided to the constructor fails with the same error.

#### Expected Output

I'd expect an empty series (if the dense `Series` has a name, or columns is provided).

Or a useful error message, if the data type is not accepted.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.4.final.0
python-bits: 64
OS: Linux
OS-release: 4.14.13-300.fc27.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_GB.utf8
LOCALE: en_GB.UTF-8

pandas: 0.23.0.dev0+152.g2952fbd55.dirty
pytest: 3.3.0
pip: 9.0.1
setuptools: 36.5.0.post20170921
Cython: 0.27.3
numpy: 1.13.3
scipy: 1.0.0
pyarrow: 0.7.1
xarray: None
IPython: 6.2.1
sphinx: 1.6.3
patsy: 0.4.1
dateutil: 2.6.1
pytz: 2017.3
blosc: None
bottleneck: 1.2.1
tables: 3.4.2
numexpr: 2.6.4
feather: None
matplotlib: 2.1.1
openpyxl: 2.4.9
xlrd: 1.1.0
xlwt: 1.3.0
xlsxwriter: 1.0.2
lxml: 4.1.1
bs4: 4.6.0
html5lib: 1.0.1
sqlalchemy: 1.1.13
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by datapythonista**
- [X] closes #19374
- [X] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

### CODE REVIEW

File: pandas/tests/sparse/frame/test_frame.py [hunk 1]
```diff
@@ -199,6 +199,31 @@ def test_constructor_from_series(self):
...
+
+        # No column name available
+        x = Series(np.random.randn(10000))
+        assert isinstance(x, Series)
+        pytest.raises(ValueError, SparseDataFrame, x)
+
+    def test_constructor_from_unknown_type(self):
+        class Unknown:
+            pass
+        pytest.raises(TypeError, SparseDataFrame, Unknown())
```
Comment (gfyoung) [inline pandas/tests/sparse/frame/test_frame.py]: Let's check error message too.

File: pandas/tests/sparse/frame/test_frame.py [hunk 2]
```diff
@@ -199,6 +199,31 @@ def test_constructor_from_series(self):
         # without sparse value raises error
         # df2 = SparseDataFrame([x2_sparse, y])
 
+    def test_constructor_from_dense_series(self):
```
Comment (gfyoung) [inline pandas/tests/sparse/frame/test_frame.py]: Reference issue number under all of your added tests.

**Comment by gfyoung** [comment]
@datapythonista : Looks pretty good so far.  Don't forget to add a `whatsnew` entry.

File: pandas/tests/sparse/frame/test_frame.py [hunk 3]
```diff
@@ -199,6 +199,31 @@ def test_constructor_from_series(self):
...
+        # getting the columns from the Series name
+        x = Series(np.random.randn(10000), name='a')
+        assert isinstance(x, Series)
+        df = SparseDataFrame(x)
+        assert isinstance(df, SparseDataFrame)
```
Comment (jreback) [inline pandas/tests/sparse/frame/test_frame.py]: construct an expected SDF and compare

File: pandas/tests/sparse/frame/test_frame.py [hunk 4]
```diff
@@ -199,6 +199,31 @@ def test_constructor_from_series(self):
...
+        assert df.columns == ['a']
+
+        # getting the columns from the columns argument
+        x = Series(np.random.randn(10000))
+        assert isinstance(x, Series)
+        df = SparseDataFrame(x, columns=['b'])
+        assert isinstance(df, SparseDataFrame)
+        assert df.columns == ['b']
+
+        # No column name available
```
Comment (jreback) [inline pandas/tests/sparse/frame/test_frame.py]: same for the rest

File: pandas/core/sparse/frame.py [hunk 5]
```diff
@@ -95,6 +95,13 @@ def __init__(self, data=None, index=None, columns=None, default_kind=None,
                                  dtype=dtype, copy=copy)
         elif isinstance(data, DataFrame):
             mgr = self._init_dict(data, data.index, data.columns, dtype=dtype)
+        elif isinstance(data, Series):
+            if columns is None and data.name is None:
```
Comment (jreback) [inline pandas/core/sparse/frame.py]: need a test to hit this case

**Review by jreback — CHANGES_REQUESTED** [review]
pls add a whatsnew note as well.

**Comment by datapythonista** [comment]
Thanks a lot for the comments. Sorry about the whatsnew, I added it when opening the PR, but forgot to add it to the commit.

@jreback, I didn't find a way to construct the expected `SparseDataFrame` without using the constructor that is being tested itself. But I added the comparison of the length and sum of the original series and the new `SparseDataFrame`, which I think should be enough (and it helped find a bug in my previous code).

Addressed all the other comments, let me know if you see anything else. Thanks!

File: pandas/core/sparse/frame.py [hunk 6]
```diff
@@ -95,6 +95,20 @@ def __init__(self, data=None, index=None, columns=None, default_kind=None,
...
+            if columns is None:
+                if data.name is None:
+                    raise ValueError('cannot pass a series '
+                                     'w/o a name or columns')
+                else:
+                    columns = [data.name]
+            elif len(columns) != 1:
+                raise ValueError('columns must be of length one '
+                                 'if data is of type Series')
+            mgr = self._init_dict(data.to_frame(columns[0]),
```
Comment (jreback) [inline pandas/core/sparse/frame.py]: don't construct like this, actually make dict

File: pandas/tests/sparse/frame/test_frame.py (hunk 3 above)
Comment (jreback) [reply pandas/tests/sparse/frame/test_frame.py]: construct a DataFrame and use .to_sparse() to construct an expected frame. we do not want to do all of these little checks, we already have well established comparison functions, e.g. ``tm.assert_sparse_equal`` for this

File: pandas/core/sparse/frame.py [hunk 7]
```diff
@@ -95,6 +95,20 @@ def __init__(self, data=None, index=None, columns=None, default_kind=None,
                                  dtype=dtype, copy=copy)
         elif isinstance(data, DataFrame):
             mgr = self._init_dict(data, data.index, data.columns, dtype=dtype)
+        elif isinstance(data, Series):
+            if columns is None:
+                if data.name is None:
+                    raise ValueError('cannot pass a series '
+                                     'w/o a name or columns')
```
Comment (jreback) [inline pandas/core/sparse/frame.py]: I don't think you need any of this, you can simply all ``to_manager`` with the column (in a list)

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by datapythonista** [comment]
Thanks for the feedback @jreback, I misunderstood what the columns argument had to do, and I was overcomplicating the code.

Now it simply creates the `DataFrame` with the `Series` name if it has one, or it uses 0 as the column if it doesn't. This seems to be consistent with the `Series` constructor and with creating a `SparseDataFrame` from a `DataFrame`, which ignores unknown columns.

I'm not 100% sure I understood what you mean by "actually contruct dic", please let me know if this new version doesn't address this too.

File: pandas/tests/sparse/frame/test_frame.py [hunk 8]
```diff
@@ -199,6 +199,33 @@ def test_constructor_from_series(self):
         # without sparse value raises error
         # df2 = SparseDataFrame([x2_sparse, y])
 
+    def test_constructor_from_dense_series(self):
+        # GH 19393
+        # series with name
+        x = Series(np.random.randn(10000), name='a')
+        assert isinstance(x, Series)
```
Comment (jreback) [inline pandas/tests/sparse/frame/test_frame.py]: you don't need this assert

File: pandas/tests/sparse/frame/test_frame.py [hunk 9]
```diff
@@ -199,6 +199,33 @@ def test_constructor_from_series(self):
...
+        res = SparseDataFrame(x)
```
Comment (jreback) [inline pandas/tests/sparse/frame/test_frame.py]: I prefer result and expected

File: pandas/tests/sparse/frame/test_frame.py [hunk 10]
```diff
@@ -199,6 +199,33 @@ def test_constructor_from_series(self):
...
+        assert isinstance(res, SparseDataFrame)
```
Comment (jreback) [inline pandas/tests/sparse/frame/test_frame.py]: you don't need the assert

File: pandas/tests/sparse/frame/test_frame.py [hunk 11]
```diff
@@ -199,6 +199,33 @@ def test_constructor_from_series(self):
...
+        exp = x.to_frame().to_sparse()
+        tm.assert_sp_frame_equal(res, exp)
+
+        # series with no name
+        x = Series(np.random.randn(10000))
+        assert isinstance(x, Series)
+        res = SparseDataFrame(x)
```
Comment (jreback) [inline pandas/tests/sparse/frame/test_frame.py]: same w.r.t naming & asserts

**Review by jreback — CHANGES_REQUESTED** [review]
small edits, other lgtm. ping on green.

**Comment by datapythonista** [comment]
Thanks once again for the comments @jreback, seems that I followed the conventions of an old test. Addressed your comments, should be all right now.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @datapythonista

**Pull request merged**